### PR TITLE
Switch to `DevWorkspace` ( devfile 2.0 ) API

### PR DESCRIPTION
**Pull request by davidfestal**
### What does this PR do?
It's under 🚧 since it should not be merged before the corresponding PRs are merged into OpenShift Console.

This PR implements the switch to use the [`DevWorkspace` (devfile 2.0)](https://github.com/devfile/kubernetes-api) API in the Che Workspace controller.

This covers only the basic use-cases that were supported with the devfile 1.0-based custom resources until now, which is also sufficient for the Cloud Shell support use-case.

New features of the DevWorkspace / devfile 2.0 are not part of this PR, and will be addressed in a distinct PR, mainly:
- Parent support
- Plugins being devfile fragments that can be overridden through a `kustomize`-like syntax.

### What issues does this PR fix or reference?

This PR references issue https://github.com/eclipse/che/issues/16626

### Is it tested? How?

Yes, it has been tested with the following docker image for the controller: `quay.io/dfestal/che-workspace-controller:use-devworkspace-api`

From the `use-devfile-2.0` branch, just set this image as the expected image, and deploy the operator again:

``` bash
export IMG=quay.io/dfestal/che-workspace-controller:use-devworkspace-api
make deploy
```
Then you should be able to create and start of of the following workspaces:

#### Theia Workspace

```yaml
kind: "DevWorkspace"
apiVersion: "devfile.io/v1alpha1"
metadata:
  "name": "my-workspace"
spec:
  started: true
  routingClass: 'openshift-oauth'
  template:
    projects:
      - name: project   
        git:
          location: "https://github.com/che-samples/web-nodejs-sample.git"
    components:
      - plugin:
          id: eclipse/che-theia/latest
      - plugin:
          id: eclipse/che-machine-exec-plugin/latest
      - plugin:
          name: "typescript-plugin"
          id: che-incubator/typescript/latest
          components:
            - container:
                name: ""
                memoryLimit: 512Mi
      - container:
          name: nodejs
          image: quay.io/eclipse/che-nodejs10-ubi:nightly
          memoryLimit: 512Mi
          endpoints:
            - name: nodejs
              configuration:
                  protocol: tcp
                  scheme: http
              targetPort: 3000
          mountSources: true
    commands:
      - exec:
          id: download dependencies
          component: nodejs
          commandLine: npm install
          workingDir: ${PROJECTS_ROOT}/project/app
      - exec:
          id: run the app
          component: nodejs
          commandLine: nodemon app.js
          workingDir: ${PROJECTS_ROOT}/project/app
      - exec:
          id: run the app (debugging enabled)
          component: nodejs
          commandLine: nodemon --inspect app.js
          workingDir: ${PROJECTS_ROOT}/project/app
      - exec:
          id: stop the app
          component: nodejs
          commandLine: >-
              node_server_pids=$(pgrep -fx '.*nodemon (--inspect )?app.js' | tr "\\n" " ") &&
              echo "Stopping node server with PIDs: ${node_server_pids}" && 
              kill -15 ${node_server_pids} &>/dev/null && echo 'Done.'
      - vscodeLaunch:
          id: Attach remote debugger
          inlined: |
            {
              "version": "0.2.0",
              "configurations": [
                {
                  "type": "node",
                  "request": "attach",
                  "name": "Attach to Remote",
                  "address": "localhost",
                  "port": 9229,
                  "localRoot": "${workspaceFolder}",
                  "remoteRoot": "${workspaceFolder}"
                }
              ]
            }
```

#### Cloud shell workspace
```yaml
kind: "DevWorkspace"
apiVersion: "devfile.io/v1alpha1"
metadata:
  "name": "openshift-terminal"
spec:
  started: false
  routingClass: 'openshift-terminal'
  template:
    components:
      - plugin:
          name: command-line-terminal
          id: che-incubator/command-line-terminal/4.5.0
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -169,6 +172,22 @@ update_crds: _set_ctx _update_crds _reset_ctx
 ### uninstall: remove namespace and all CRDs from cluster
 uninstall: _set_ctx _do_uninstall _reset_ctx
 
+update_devworkspace_api:
+	go mod edit --require github.com/devfile/kubernetes-api@$(DEVWORKSPACE_API_VERSION)
+	go mod download
+	go mod vendor
+
+### ...
+update_devworkspace_crds:
```
Comment (amisevsk) [inline Makefile:181]: The structure for the doc comments is
```
### <name of makefile rule>: <explanation of what it does
```
This is used to get the output of `make help`.
```suggestion
### update_devworkspace_api: update version of devworkspace_api in go.mod
update_devworkspace_api:
	go mod edit --require github.com/devfile/kubernetes-api@$(DEVWORKSPACE_API_VERSION)
	go mod download
	go mod vendor

### update_devworkspace_crds: download updated devworkspace_crd definitions
update_devworkspace_crds:
```
You will also need to change line 234 below to increase column width.
```
@sed -n 's/^### /    /p' $< | awk 'BEGIN { FS=":" } { printf "%-22s -%s\n", $$1, $$2 }'

to

@sed -n 's/^### /    /p' $< | awk 'BEGIN { FS=":" } { printf "%-30s -%s\n", $$1, $$2 }'
```

File: go.mod [hunk 2]
```diff
@@ -5,16 +5,19 @@ go 1.12
 // Che Plugin Broker branch devworkspace-controller
 require github.com/eclipse/che-plugin-broker v3.1.1-0.20200207223144-b20597f15e4c+incompatible
 
+// use-devfile-2.0-in-workspace-controller
+require github.com/devfile/kubernetes-api v0.0.0-20200603111415-8e1485401159
+
+// replace github.com/devfile/kubernetes-api => /home/dfestal/go/src/github.com/devfile/kubernetes-api
```
Comment (amisevsk) [inline go.mod:11]: Accidental inclusion? 

It's possible to point at a branch pushed remotely, which we do e.g. for the plugin broker above.

File: go.mod [hunk 3]
```diff
@@ -5,16 +5,19 @@ go 1.12
...
 // Operator Framework 0.17.x
 require (
 	github.com/eclipse/che-go-jsonrpc v0.0.0-20181205102516-87cdb8da2597 // indirect
 	github.com/go-logr/logr v0.1.0
-	github.com/go-openapi/spec v0.19.4
 	github.com/google/go-cmp v0.4.0
 	github.com/google/uuid v1.1.1
 	github.com/openshift/api v3.9.1-0.20190424152011-77b8897ec79a+incompatible
 	gopkg.in/yaml.v2 v2.2.8
-	k8s.io/kube-openapi v0.0.0-20191107075043-30be4d16710a
```
Comment (amisevsk) [inline go.mod:17]: I get a lot of local changes when I run `go mod tidy`

**Comment by amisevsk** [comment]
Rebased on master.

@davidfestal I squashed your changes into two commits (separating out autogenerated file changes into a separate commit) to make the rebase easier. LMK if that is okay with you.

File: Makefile [hunk 4]
```diff
@@ -8,6 +8,7 @@ WEBHOOK_ENABLED ?= false
 DEFAULT_ROUTING ?= basic
 ADMIN_CTX ?= ""
 REGISTRY_ENABLED ?= true
+DEVWORKSPACE_API_VERSION ?= master
```
[the remark this answers is not in the document]
Comment (davidfestal) [reply Makefile:11]: Sure, that was the initial idea, and then I forgot doing it.

File: Makefile [hunk 5]
```diff
@@ -171,6 +176,23 @@ update_crds: _set_ctx _update_crds _reset_ctx
 ### uninstall: remove namespace and all CRDs from cluster
 uninstall: _set_ctx _do_uninstall _reset_ctx
 
+### update_devworkspace_api: update version of devworkspace crds in go.mod
+update_devworkspace_api:
+	go mod edit --require github.com/devfile/kubernetes-api@$(DEVWORKSPACE_API_VERSION)
+	go mod download
+	go mod vendor
```
[the remark this answers is not in the document]
Comment (davidfestal) [reply Makefile]: Oh, OK, I didn't know that.

**Comment by davidfestal** [comment]
I'm currently reviewing the last changes.

**Comment by davidfestal** [comment]
Everything's OK for me, apart from this https://github.com/che-incubator/che-workspace-operator/pull/92#discussion_r435829124 that should probably be done before merging.

I assume that we would manage the last renaming of apiGroups for another PRs:

- kind: `DevWorkspace` apiGroup: `workspace.devfile.io`
- kind: `Component` apiGroup: `component.devfile.io`
- kind: `WorkspaceRouting` apiGroup: `routing.devfile.io`

File: samples/theia-nodejs.yaml [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+      - plugin:
+          id: eclipse/che-theia/latest
+      - plugin:
+          id: eclipse/che-machine-exec-plugin/latest
+      - plugin:
+          name: "typescript-plugin"
+          id: che-incubator/typescript/latest
+          components:
+            - container:
+                name: ""
```
[the remark this answers is not in the document]
Comment (davidfestal) [reply samples/theia-nodejs.yaml:23]: My thought was that explicitly empty container names would override the containers in the order of overrides. => first container override with empty name will override the config of the first container component.
But you're right that another interesting use-case would be to override all cotainer components, and in this case we could probably dedicate the `*` name.

**Comment by davidfestal** [comment]
> Everything's OK for me, apart from this [#92 (comment)](https://github.com/che-incubator/che-workspace-operator/pull/92#discussion_r435829124) that should probably be done before merging.

@sleshchenko @amisevsk Do you want me to do it, or does one of you plan to just do this modification ?

**Comment by sleshchenko** [comment]
@davidfestal It should be already implemented in https://github.com/che-incubator/che-workspace-operator/pull/92/commits/d3b7b159b0f814c312598da1e86def1ce84c2f1f
if you could test it as well - would be great.

I also tried to implement configuring a repo, like to reference my fork like to check PR, but seems that go mod and forks are not friends =)
Screenshot_20200605_151428
So, the way to test it now - push to devfile/kubernetes-api without fork. Later we will be able to return back to this problem and maybe we'll find some solution

File: Makefile (hunk 4 above)
[the remark this answers is not in the document]
Comment (sleshchenko) [reply Makefile:11]: Should be fixed in https://github.com/che-incubator/che-workspace-operator/commit/d3b7b159b0f814c312598da1e86def1ce84c2f1f

**Comment by amisevsk** [comment]
> I also tried to implement configuring a repo, like to check devworkspace-api PR from fork, but seems that go mod and forks are not friends =)

I believe you could do this via a `replace` directive

**Comment by amisevsk** [comment]
Rebased on master to pick up https://github.com/che-incubator/che-workspace-operator/pull/97

**Pull request merged**